### PR TITLE
[#27] QHull refactoring

### PR DESCRIPTION
**Pull request by ilia-glushchenko**

### CODE REVIEW

File: Epona/include/Epona/Analysis.hpp [hunk 1]
```diff
@@ -1,11 +1,12 @@
 /*
-* Copyright (C) 2017 by Godlike
+* Copyright (C) 2019 by Godlike
```
Comment (MrDmitry) [inline Epona/include/Epona/Analysis.hpp]: living in the future :open_mouth:

File: Epona/include/Epona/Analysis.hpp [hunk 2]
```diff
@@ -373,6 +374,19 @@ float LineSegmentPointDistance(
...
  */
 glm::vec3 CalculateBarycentricCoordinates(glm::vec3 p, glm::vec3 a, glm::vec3 b, glm::vec3 c);
+
+/*
+ * @brief Calculate area of the triangle and returns true if area is equal to zero
+ *
+ * @param a triangle's point
+ * @param b triangle's point
+ * @param c triangle's point
+ * @return if points are on the same line
```
Comment (MrDmitry) [inline Epona/include/Epona/Analysis.hpp]:
```
@return true if points are on the same line
```
?

File: Epona/include/Epona/Analysis.hpp [hunk 3]
```diff
@@ -373,6 +374,19 @@ float LineSegmentPointDistance(
  * @return barycentric coordinates
  */
 glm::vec3 CalculateBarycentricCoordinates(glm::vec3 p, glm::vec3 a, glm::vec3 b, glm::vec3 c);
+
+/*
+ * @brief Calculate area of the triangle and returns true if area is equal to zero
```
Comment (MrDmitry) [inline Epona/include/Epona/Analysis.hpp:379]:
```
returns true if area is equal to zero
```
sounds different from
```
points are on the same line
```

Please choose 1 form of expressing intent

File: Epona/include/Epona/Debug.hpp [hunk 4]
```diff
@@ -1,5 +1,5 @@
 /*
-* Copyright (C) 2018 by Godlike
+* Copyright (C) 2019 by Godlike
```
Comment (MrDmitry) [inline Epona/include/Epona/Debug.hpp]: why this isolated change? please revert if nothing in file is changed - such changes introduce misleading `git blame` and `git log` behavior

File: Epona/include/Epona/FloatingPoint.hpp [hunk 5]
```diff
@@ -1,5 +1,5 @@
 /*
-* Copyright (C) 2017 by Godlike
+* Copyright (C) 2019 by Godlike
```
Comment (MrDmitry) [inline Epona/include/Epona/FloatingPoint.hpp]: please revert

File: Epona/include/Epona/HalfEdgeDataStructure.hpp [hunk 6]
```diff
@@ -321,6 +355,9 @@ class HalfEdgeDataStructure
         */
         const_face_iterator GetAdjacentFaceIterator() const;
 
+        //! Stores face index
```
Comment (MrDmitry) [inline Epona/include/Epona/HalfEdgeDataStructure.hpp]: `Stores` is redundant

File: Epona/include/Epona/HalfEdgeDataStructure.hpp [hunk 7]
```diff
@@ -366,7 +403,42 @@ class HalfEdgeDataStructure
...
+    {
+        MakeFace(a, b, c, 0);
+    }
+
+    /**
+     * @brief Inserts face into the current data structure
+     * @param a vertex index
+     * @param b vertex index
+     * @param c vertex index
+     * @param hp hyperplane containing input vertices
```
Comment (MrDmitry) [inline Epona/include/Epona/HalfEdgeDataStructure.hpp]: there's no `hp` parameter

File: Epona/include/Epona/HalfEdgeDataStructure.hpp [hunk 8]
```diff
@@ -366,7 +403,42 @@ class HalfEdgeDataStructure
...
+     * @param index outside face index
+     */
+    void MakeFace(uint64_t a, uint64_t b, uint64_t c, uint32_t index);
+
+    /**
+     * @brief Inserts face into the current data structure
```
Comment (MrDmitry) [inline Epona/include/Epona/HalfEdgeDataStructure.hpp]: Returns a face iterator

File: Epona/include/Epona/HalfEdgeDataStructure.hpp [hunk 9]
```diff
@@ -366,7 +403,42 @@ class HalfEdgeDataStructure
...
+     * @param   index   face indices array of size 3
+     */
+    template < typename T >
+    decltype(auto) GetFace(T index)
+    {
+        return GetFace(index[0], index[1], index[2]);
+    }
+
+    /**
+     * @brief Inserts face into the current data structure
```
Comment (MrDmitry) [inline Epona/include/Epona/HalfEdgeDataStructure.hpp]: Returns a face iterator

File: Epona/include/Epona/HyperPlane.hpp [hunk 10]
```diff
@@ -1,5 +1,5 @@
 /*
-* Copyright (C) 2017 by Godlike
+* Copyright (C) 2019 by Godlike
```
Comment (MrDmitry) [inline Epona/include/Epona/HyperPlane.hpp]: please revert

File: Epona/include/Epona/JacobiEigenvalue.hpp [hunk 11]
```diff
@@ -1,83 +1,184 @@
...
+#include <tuple>
+
+namespace
+{
+/**
+ * @brief Finds maximal absolute value off diagonal matrix element and sets its indices
+ *
+ * @param[in] mat matrix to search
+ * @param[out] i max element row index
+ * @param[out] j max element column index
```
Comment (MrDmitry) [inline Epona/include/Epona/JacobiEigenvalue.hpp]: there are no `[out]` params, instead there is a `@return` tuple

File: Epona/include/Epona/JacobiEigenvalue.hpp [hunk 12]
```diff
@@ -1,83 +1,184 @@
...
 #include <Epona/FloatingPoint.hpp>
+#define GLM_ENABLE_EXPERIMENTAL
+#include <glm/gtx/norm.hpp>
 #include <glm/glm.hpp>
+#include <tuple>
+
+namespace
+{
+/**
+ * @brief Finds maximal absolute value off diagonal matrix element and sets its indices
```
Comment (MrDmitry) [inline Epona/include/Epona/JacobiEigenvalue.hpp]:
```
and returns its indices
```

File: Epona/include/Epona/JacobiEigenvalue.hpp [hunk 13]
```diff
@@ -1,83 +1,184 @@
...
+    return { i, j };
+}
+
+/**
+ * @brief Calculates rotation angle for a given matrix and its element
+ *
+ * @param[in] mat matrix to rotate
+ * @param[in] i element row index
+ * @param[in] j element column index
+ *
```
Comment (MrDmitry) [inline Epona/include/Epona/JacobiEigenvalue.hpp:53]:
```
* @param[in] coverageThreshold coverage precision threshold
```

File: Epona/include/Epona/JacobiEigenvalue.hpp [hunk 14]
```diff
@@ -1,83 +1,184 @@
...
 
 namespace epona
 {
+
+/**
+ * @brief Calculates eigenvalues and eigenvectors
+ *
+ * @param[in] symmetricMatrix target symmetric matrix
+ * @param[in] coverageThreshold coverage precision threshold
+ * @param[in] maxIterations maximum amount of iteration
```
Comment (MrDmitry) [inline Epona/include/Epona/JacobiEigenvalue.hpp]: `maximum number of iterations`

File: Epona/include/Epona/JacobiEigenvalue.hpp [hunk 15]
```diff
@@ -1,83 +1,184 @@
...
+
+    return std::tuple<glm::mat3, glm::vec3>{ eigenvectors,  glm::vec3{ symmetricMatrix[0][0], symmetricMatrix[1][1], symmetricMatrix[2][2] } };
+}
+
+/**
+ * @brief Calculates eigenvalues
+ *
+ * @param[in] symmetricMatrix target symmetric matrix
+ * @param[in] coverageThreshold coverage precision threshold
+ * @param[in] maxIterations maximum amount of iteration
```
Comment (MrDmitry) [inline Epona/include/Epona/JacobiEigenvalue.hpp]: `maximum number of iterations`

File: Epona/include/Epona/JacobiEigenvalue.hpp [hunk 16]
```diff
@@ -1,83 +1,184 @@
...
+    return { symmetricMatrix[0][0], symmetricMatrix[1][1], symmetricMatrix[2][2] };
+}
+
 /**
- * @brief Jacobi eigenvalue calculation algorithm
+ * @brief Calculates eigenvectors
+ *
+ * @param[in] symmetricMatrix target symmetric matrix
+ * @param[in] coverageThreshold coverage precision threshold
+ * @param[in] maxIterations maximum amount of iteration
```
Comment (MrDmitry) [inline Epona/include/Epona/JacobiEigenvalue.hpp]: `maximum number of iterations`

File: Epona/include/Epona/QuickhullConvexHull.hpp [hunk 17]
```diff
@@ -25,512 +25,395 @@
...
+        return aVertex == index || bVertex == index;
+    }
+};
+
 /**
- * @brief Quickhull convex hull calculation algorithm
- * @tparam VerticesType STL compatible random access glm::vec3 container
+ * @brief Calculates initial tetrehedron for the convex hull
+ *
+ * @note vertices size must be >= 4 and non degenerate
```
Comment (MrDmitry) [inline Epona/include/Epona/QuickhullConvexHull.hpp]: `nondegenerate`

File: Epona/include/Epona/QuickhullConvexHull.hpp [hunk 18]
```diff
@@ -25,512 +25,395 @@
...
+ *
+ * @param vertices input vertex data
```
Comment (MrDmitry) [inline Epona/include/Epona/QuickhullConvexHull.hpp:78]: also add `@return`

File: Epona/include/Epona/QuickhullConvexHull.hpp [hunk 19]
```diff
@@ -25,512 +25,395 @@
...
-public:
-    class Face;
-    using Faces = std::list<Face>;
-    using Vertices = VerticesType;
-
-    /**
-     * @brief Convex hull face container
-     */
-    class Face
+    //Find directions of the maximum spread for the calculation of the frist approximation of the convex hull
```
Comment (MrDmitry) [inline Epona/include/Epona/QuickhullConvexHull.hpp]: `first`

File: Epona/include/Epona/QuickhullConvexHull.hpp [hunk 20]
```diff
@@ -25,512 +25,395 @@
...
-    bool AddVertex(size_t index)
-    {
-        //Create stack of convex hull faces
-        std::list<typename Faces::iterator> faceIteratorStack;
-        for (auto faceIt = m_faces.begin(); faceIt != m_faces.end(); ++faceIt)
+namespace epona
+{
+
+/**
+ * @brief Updates conex hull based on the vertices data
```
Comment (MrDmitry) [inline Epona/include/Epona/QuickhullConvexHull.hpp]: `convex`

File: Epona/include/Epona/QuickhullConvexHull.hpp [hunk 21]
```diff
@@ -25,512 +25,395 @@
...
+ *
+ * @note vertices size must be >= 4 and non degenerate
```
Comment (MrDmitry) [inline Epona/include/Epona/QuickhullConvexHull.hpp]: `nondegenerate`

File: Epona/include/Epona/QuickhullConvexHull.hpp [hunk 22]
```diff
@@ -25,512 +25,395 @@
...
+ * @attention The only way to update vertex data is to append some additional vertices
+ *
+ * @param[in, out] hull convex hull to update
+ * @param[in] vertices updated input vertex data on which hull was based
+ * @param[in] maxIterations maximum allowed iterations count
```
Comment (MrDmitry) [inline Epona/include/Epona/QuickhullConvexHull.hpp]: `maximum number of iterations` to follow the same phrasing?

File: Epona/include/Epona/QuickhullConvexHull.hpp [hunk 23]
```diff
@@ -25,512 +25,395 @@
...
+ * @param[in] distanceThreshold vertex is an inlier if its signed distance is less or equal to the threshold
+ */
```
Comment (MrDmitry) [inline Epona/include/Epona/QuickhullConvexHull.hpp:188]: add `@return`

File: Epona/include/Epona/QuickhullConvexHull.hpp [hunk 24]
```diff
@@ -25,512 +25,395 @@
...
+                std::vector<uint32_t> visitedFaces(1, f);
+                visitedFaces.reserve(hull.faces.size() / 2);
+                glm::vec3 const extremeVertex = vertices[faceOutliers.extremeIndices[f]];
+
+                for (uint32_t vf = 0; vf < visibleFaces.size(); ++vf)
+                {
+                    auto adjFaceIt = hull.heds.GetFace(hull.faces[visibleFaces[vf]])->GetAdjacentFaceIterator();
+
+                    for (uint8_t i = 0; i < 3; ++i, ++adjFaceIt) {
+                        //NOTE: will work faster then open address map till ~ visitedFaces.size() < 100
```
Comment (MrDmitry) [inline Epona/include/Epona/QuickhullConvexHull.hpp]: `than`

File: Epona/include/Epona/QuickhullConvexHull.hpp [hunk 25]
```diff
@@ -25,512 +25,395 @@
...
+ * @param[in] distanceThreshold vertex is an inlier if its signed distance is less or equal to the threshold
+ */
+inline bool RecalculateConvexHull(ConvexHull& hull, std::vector<glm::vec3> const& vertices, uint32_t maxIterations = 100, float distanceThreshold = 1e-3f)
+{
+    bool changed = false;
+
+    std::vector<uint32_t> indices;
+    indices.reserve(vertices.size());
+    for (uint32_t v = 0; v < vertices.size(); ++v)
+        if (std::find(hull.indices.begin(), hull.indices.end(), v) == hull.indices.end())
```
Comment (MrDmitry) [inline Epona/include/Epona/QuickhullConvexHull.hpp:196]: `std::none_of` instead?

File: Epona/include/Epona/QuickhullConvexHull.hpp [hunk 26]
```diff
@@ -25,512 +25,395 @@
...
+                visitedFaces.reserve(hull.faces.size() / 2);
+                glm::vec3 const extremeVertex = vertices[faceOutliers.extremeIndices[f]];
+
+                for (uint32_t vf = 0; vf < visibleFaces.size(); ++vf)
+                {
+                    auto adjFaceIt = hull.heds.GetFace(hull.faces[visibleFaces[vf]])->GetAdjacentFaceIterator();
+
+                    for (uint8_t i = 0; i < 3; ++i, ++adjFaceIt) {
+                        //NOTE: will work faster then open address map till ~ visitedFaces.size() < 100
+                        if (std::find(visitedFaces.begin(), visitedFaces.end(), adjFaceIt->index) == visitedFaces.end())
```
Comment (MrDmitry) [inline Epona/include/Epona/QuickhullConvexHull.hpp:253]: `std::none_of` instead?

File: Epona/include/Epona/QuickhullConvexHull.hpp [hunk 27]
```diff
@@ -25,512 +25,395 @@
...
-
-                //If face is unvisited
-                if (visitedFaces.find(&*adjFace) == visitedFaces.end()
-                    && visibleFacesSet.find(&*adjFace) == visibleFacesSet.end())
+                auto adjFaceIt = hull.heds.GetFace(hull.faces[vf])->GetAdjacentFaceIterator();
+                for (uint8_t i = 0; i < 3; ++i, ++adjFaceIt)
                 {
-                    //If face is visible add it to the visible faces set
-                    if (adjFace->GetHyperPlane().SignedDistance(m_vertexBuffer[extremalVertexIndex]) > 0)
+                    if (std::find(visibleFaces.begin(), visibleFaces.end(), adjFaceIt->index) == visibleFaces.end())
```
Comment (MrDmitry) [inline Epona/include/Epona/QuickhullConvexHull.hpp:286]: `std::none_of` instead?

File: Epona/include/Epona/QuickhullConvexHull.hpp [hunk 28]
```diff
@@ -25,512 +25,395 @@
...
                     {
-                        visibleFaces.push_back(adjFace);
-                        visibleFacesSet.insert(&*adjFace);
-
-                        std::vector<size_t>& faceVertices = adjFace->GetVertices();
-                        partitionMarkedVertexIndices.insert(partitionMarkedVertexIndices.end(), faceVertices.begin(), faceVertices.end());
+                        auto halfEdge = static_cast<HalfEdgeDataStructure::HalfEdge*>(adjFaceIt);
+                        ::Ridge const ridge{ static_cast<uint32_t>(halfEdge->vertexIndex), static_cast<uint32_t>(halfEdge->prev->vertexIndex) };
+                        //NOTE: will work faster then open address map till ~ boundary.size() < 100
```
Comment (MrDmitry) [inline Epona/include/Epona/QuickhullConvexHull.hpp]: `than`

File: Epona/include/Epona/QuickhullConvexHull.hpp [hunk 29]
```diff
@@ -25,512 +25,395 @@
...
+                        if (std::find(boundary.begin(), boundary.end(), ridge) == boundary.end())
```
Comment (MrDmitry) [inline Epona/include/Epona/QuickhullConvexHull.hpp:291]: `std::none_of` instead?

File: Epona/include/Epona/QuickhullConvexHull.hpp [hunk 30]
```diff
@@ -25,512 +25,395 @@
...
+                        {
+                            boundary.push_back(ridge);
+                        }
                     }
-                    //If face is not visible then find a ridge and save it
-                    else
+                }
+            }
+            //FixMe: Checks for degenerate case, that should be prevented by threshold check, is there a way to make it better?
+            if (std::find(boundary.begin(), boundary.end(), extremeOutlierIndex) != boundary.end())
```
Comment (MrDmitry) [inline Epona/include/Epona/QuickhullConvexHull.hpp:299]: `std::any_of` instead?

File: Epona/include/Epona/QuickhullConvexHull.hpp [hunk 31]
```diff
@@ -25,512 +25,395 @@
...
+                return true;
+
+            //Collect inliers as outliers and remove them from hull indices
+            temporaryOutliers.clear();
+            temporaryOutliers.reserve(visibleFaces.size() * 3 + newOutliersCount);
+            for (uint32_t vf : visibleFaces) {
+                auto face = hull.faces[vf];
+                for (uint8_t i = 0; i < 3; ++i) {
+                    //NOTE: will work faster then open address map till ~ (boundary/outliers).size() < 100
```
Comment (MrDmitry) [inline Epona/include/Epona/QuickhullConvexHull.hpp]: `than`

File: Epona/include/Epona/QuickhullConvexHull.hpp [hunk 32]
```diff
@@ -25,512 +25,395 @@
...
+                    if (std::find(boundary.begin(), boundary.end(), face[i]) == boundary.end()
+                        && std::find(temporaryOutliers.begin(), temporaryOutliers.end(), face[i]) == temporaryOutliers.end())
```
Comment (MrDmitry) [inline Epona/include/Epona/QuickhullConvexHull.hpp:310]: `std::none_of` instead?

File: Epona/include/Epona/QuickhullConvexHull.hpp [hunk 33]
```diff
@@ -25,512 +25,395 @@
...
-                        ridgeIndices.reserve(2);
-
-                        auto adjFaceIndices = adjFace->GetIndices();
-                        auto faceIndices = (*visibleFaceIt)->GetIndices();
-                        std::set_intersection(adjFaceIndices.begin(), adjFaceIndices.end(),
-                            faceIndices.begin(), faceIndices.end(), std::back_inserter(ridgeIndices));
-
-                        horizonRidges.emplace_back(ridgeIndices[0], ridgeIndices[1]);
+                        temporaryOutliers.push_back(face[i]);
+                        //NOTE: will work faster then open address map till ~ hull.indieces.size() < 100
```
Comment (MrDmitry) [inline Epona/include/Epona/QuickhullConvexHull.hpp]: `than`

File: Epona/include/Epona/QuickhullConvexHull.hpp [hunk 34]
```diff
@@ -25,512 +25,395 @@
...
+                    faceOutliers.extremeSignedDistances[faceIndex] = -FLT_MAX;
+                    faceOutliers.outliers[faceIndex].clear();
+                }
+                else
+                {
+                    hull.faces.emplace_back(extremeOutlierIndex, ridge.aVertex, ridge.bVertex);
+                    hull.planes.emplace_back(vertices[extremeOutlierIndex], vertices[ridge.aVertex], vertices[ridge.bVertex], &mean);
+                }
+                hull.heds.MakeFace(extremeOutlierIndex, ridge.aVertex, ridge.bVertex, faceIndex);
+                //NOTE: will work faster then open address map till ~ hull.indieces.size() < 100
```
Comment (MrDmitry) [inline Epona/include/Epona/QuickhullConvexHull.hpp]: `than`

File: Epona/include/Epona/QuickhullConvexHull.hpp [hunk 35]
```diff
@@ -25,512 +25,395 @@
...
+                if (std::find(hull.indices.begin(), hull.indices.end(), extremeOutlierIndex) == hull.indices.end())
```
Comment (MrDmitry) [inline Epona/include/Epona/QuickhullConvexHull.hpp:355]: `std::none_of` instead?

File: Epona/include/Epona/QuickhullConvexHull.hpp [hunk 36]
```diff
@@ -25,512 +25,395 @@
...
-    void Refine()
-    {
-        //Create stack of convex hull faces
-        std::list<typename Faces::iterator> faceStack;
-        for (auto faceIt = m_faces.begin(); faceIt != m_faces.end(); ++faceIt)
-        {
-            faceStack.push_back(faceIt);
-        }
+/**
+ * @brief Calculates conex hull based on the vertices data using Quickhull convex hull algorithm
```
Comment (MrDmitry) [inline Epona/include/Epona/QuickhullConvexHull.hpp]: `convex`

File: Epona/include/Epona/QuickhullConvexHull.hpp [hunk 37]
```diff
@@ -25,512 +25,395 @@
...
+ *
+ * @note vertices size must be >= 4 and non degenerate
+ *
+ * @param[in] vertices updated input vertex data on which hull was based
+ * @param[in] maxIterations maximum allowed iterations count
```
Comment (MrDmitry) [inline Epona/include/Epona/QuickhullConvexHull.hpp]: `maximum number of iterations` to follow the same phrasing?

File: Epona/include/Epona/QuickhullConvexHull.hpp [hunk 38]
```diff
@@ -25,512 +25,395 @@
...
+ * @param[in] distanceThreshold vertex is an inlier if its signed distance is less or equal to the threshold
```
Comment (MrDmitry) [inline Epona/include/Epona/QuickhullConvexHull.hpp]: sounds a bit wordy

File: Epona/include/Epona/QuickhullConvexHull.hpp [hunk 39]
```diff
@@ -25,512 +25,395 @@
...
+/**
+ * @brief Updates conex hull based on the vertices data
+ *
+ * @note vertices size must be >= 4 and non degenerate
+ * @attention The only way to update vertex data is to append some additional vertices
+ *
+ * @param[in, out] hull convex hull to update
+ * @param[in] vertices updated input vertex data on which hull was based
+ * @param[in] maxIterations maximum allowed iterations count
+ * @param[in] distanceThreshold vertex is an inlier if its signed distance is less or equal to the threshold
```
Comment (MrDmitry) [inline Epona/include/Epona/QuickhullConvexHull.hpp]: sounds a bit wordy

File: Epona/include/Epona/debug/DebugDummy.hpp [hunk 40]
```diff
@@ -1,5 +1,5 @@
 /*
-* Copyright (C) 2018 by Godlike
+* Copyright (C) 2019 by Godlike
```
Comment (MrDmitry) [inline Epona/include/Epona/debug/DebugDummy.hpp]: please revert

File: Epona/include/Epona/debug/DebugImplementation.hpp [hunk 41]
```diff
@@ -31,59 +30,45 @@ class DebugImplementation
...
     * @brief Initializes QuickhullConvexHull callback instance
     *
-    * @tparam VertexBuffer convex hull vertex buffer type
-    *
     * @param callback callback function object
     */
-    template < typename BufferType >
     static void SetQuickhullConvexHullCallback(
-        std::function<void(QuickhullConvexHull<BufferType>&, BufferType&)> callback
+        std::function<void(epona::ConvexHull&, std::vector<glm::vec3>&)> callback
```
Comment (MrDmitry) [inline Epona/include/Epona/debug/DebugImplementation.hpp:49]: alias this `std::function` with `using` for simplicity?

File: Epona/sources/HalfEdgeDataStructure.cpp [hunk 42]
```diff
@@ -36,39 +36,33 @@ HalfEdgeDataStructure::Face::const_face_iterator HalfEdgeDataStructure::Face::Ge
...
             m_halfEdgeList.emplace(m_halfEdgeList.end()),
             m_halfEdgeList.emplace(m_halfEdgeList.end())
         }};
 
-        //Allocate face
+        //Allocate face and set HyperPlane
         auto const backFaceIterator = m_facesList.emplace(m_facesList.end(), *newHalfEdges.back());
         m_faceIteratorMap[&*backFaceIterator] = backFaceIterator;
         m_faceVerticesIteratorMap[faceVerticesKey] = backFaceIterator;
+		backFaceIterator->index = index;
```
Comment (MrDmitry) [inline Epona/sources/HalfEdgeDataStructure.cpp]: fix indent

File: test/AnalysisTests.cpp [hunk 43]
```diff
@@ -14,3 +17,2827 @@ TEST_CASE("Expected value", "[statistics]")
...
+		{-0.392023f,0.306074f,-0.111411f},{-0.076496f,-0.147506f,0.303444f},{0.174756f,0.134362f,0.0740491f},
+		{0.193958f,0.128851f,0.0764992f},{0.202891f,0.125372f,0.0889489f},{-0.0161249f,-0.128883f,0.364271f},
+		{-0.0108711f,-0.139282f,0.370477f},{0.00423457f,-0.13323f,0.372734f},{0.0213594f,-0.130105f,0.374708f},
+		{0.0406514f,-0.130099f,0.374599f},{0.0599434f,-0.134394f,0.377383f},{0.0792675f,-0.130028f,0.37503f},
+		{0.0985466f,-0.130182f,0.372824f},{0.116559f,-0.135539f,0.372322f},{0.137111f,-0.139237f,0.371872f},
+		{0.156403f,-0.130774f,0.363783f},{0.175573f,-0.130221f,0.358779f},{-0.184473f,0.336015f,0.0502106f},
+		{-0.461281f,-0.0544227f,0.13352f},{-0.355394f,0.323424f,-0.246577f},{-0.451744f,-0.00167199f,0.0960677f},
+		{-0.488778f,0.0948844f,0.0762548f},{0.0491334f,0.377615f,-0.128253f},{-0.290123f,0.281174f,-0.147828f},
+		{-0.316212f,0.293643f,0.0745378f},{-0.0885084f,0.00246939f,-0.107488f},{-0.245648f,0.267271f,0.188579f},
+		{-0.35522f,-0.131314f,0.277071f},{-0.3365f,-0.131655f,0.284145f},{-0.0710685f,0.0887624f,-0.0009452
```
Comment (MrDmitry) [inline test/AnalysisTests.cpp]: add empty line?

**Review by MrDmitry — CHANGES_REQUESTED** [review]

File: Epona/include/Epona/QuickhullConvexHull.hpp (hunk 26 above)
Comment (ilia-glushchenko) [reply Epona/include/Epona/QuickhullConvexHull.hpp:253]: @dsodintsov If it would work with value type and not just unary predicate, I would use it. But as it is since C++11, the predicate itself would require a lamba function or some function defined somewhere, which kind of contradicts the reason to use it in a first place.

File: Epona/include/Epona/QuickhullConvexHull.hpp (hunk 30 above)
Comment (ilia-glushchenko) [reply Epona/include/Epona/QuickhullConvexHull.hpp:299]: @dsodintsov same reasoning as `std::none_of`

File: Epona/include/Epona/debug/DebugImplementation.hpp (hunk 41 above)
Comment (ilia-glushchenko) [reply Epona/include/Epona/debug/DebugImplementation.hpp:49]: I don't really favor those anymore, I'd rather see the full expression, for clarity sake

File: Epona/include/Epona/JacobiEigenvalue.hpp (hunk 16 above)
Comment (MrDmitry) [reply Epona/include/Epona/JacobiEigenvalue.hpp]: still actual

File: Epona/include/Epona/QuickhullConvexHull.hpp (hunk 23 above)
Comment (MrDmitry) [reply Epona/include/Epona/QuickhullConvexHull.hpp:188]: still actual

File: Epona/include/Epona/FloatingPoint.hpp [hunk 44]
```diff
@@ -1,5 +1,5 @@
 /*
-* Copyright (C) 2017 by Godlike
+* Copyright (C) 2018 by Godlike
```
Comment (MrDmitry) [inline Epona/include/Epona/FloatingPoint.hpp]: revert?

File: Epona/include/Epona/HyperPlane.hpp [hunk 45]
```diff
@@ -1,5 +1,5 @@
 /*
-* Copyright (C) 2017 by Godlike
+* Copyright (C) 2018 by Godlike
```
Comment (MrDmitry) [inline Epona/include/Epona/HyperPlane.hpp]: revert?

**Review by MrDmitry — CHANGES_REQUESTED** [review]

File: Epona/include/Epona/HalfEdgeDataStructure.hpp [hunk 46]
```diff
@@ -366,7 +403,41 @@ class HalfEdgeDataStructure
...
+     * @param b vertex index
+     * @param c vertex index
+     * @param index outside face index
+     */
+    void MakeFace(uint64_t a, uint64_t b, uint64_t c, uint32_t index);
+
+    /**
+     * @brief Returns face iterator
+     * @type T array-like type with overaloded operator[]
+     * @return face iterator
```
Comment (MrDmitry) [inline Epona/include/Epona/HalfEdgeDataStructure.hpp:424]: `@param index`?

File: Epona/include/Epona/HalfEdgeDataStructure.hpp [hunk 47]
```diff
@@ -366,7 +403,41 @@ class HalfEdgeDataStructure
...
+    template < typename T >
+    decltype(auto) GetFace(T index)
+    {
+        return GetFace(index[0], index[1], index[2]);
+    }
+
+    /**
+     * @brief Returns face iterator
+     * @type T array-like type with overaloded operator[]
+     * @return face iterator
```
Comment (MrDmitry) [inline Epona/include/Epona/HalfEdgeDataStructure.hpp:436]: `@param index`?

**Review by MrDmitry — CHANGES_REQUESTED** [review]

File: Epona/include/Epona/FloatingPoint.hpp (hunk 44 above)
Comment (ilia-glushchenko) [reply Epona/include/Epona/FloatingPoint.hpp]: This file was changed in 2018

File: Epona/include/Epona/HyperPlane.hpp (hunk 45 above)
Comment (ilia-glushchenko) [reply Epona/include/Epona/HyperPlane.hpp]: This file was changed in 2018

**Review by MrDmitry — APPROVED** [review]
